feat(aws/cloudformation): Add cloudformation caching agent and provider

xavileon opened the pull request:
This patch adds two components necessary to specify and implement a
cloud formation stage in a pipeline.

* Cloudformation caching agent: it's responsible to periodically query
aws to cache the relevant information from a cloudformation stack,
including the status whenever it's in an errored state.

* Cloudformation provider: it provides an API to query the cache and get
a cloudformation stack given a stackId or list existing stacks for an
aws account and (optionally) a region.

The cloudformation controller that will be used by Orca as well as the
task that deploys this cloud formation will be added in separate PRs.

This is the first PR of a series to address https://github.com/spinnaker/spinnaker/issues/736

Signed-off-by: Xavi León <xavi.leon@schibsted.com>

xavileon commented:
Regarding this comment https://github.com/spinnaker/clouddriver/pull/3278#issuecomment-454773226 All the code in the `clouddriver-aws`module is in groovy and added the new code in groovy for consistency, is it preferable to move it to Java in any case?

ethanfrogers commented:
@xavileon yes, any new code should be written in something non-Groovy, however if there is already another language being used in the module (Kotlin, Java, etc) you should use that. I believe we try to keep modules to no more than 2 languages.

xavileon commented:
> @xavileon yes, any new code should be written in something non-Groovy, however if there is already another language being used in the module (Kotlin, Java, etc) you should use that. I believe we try to keep modules to no more than 2 languages.

ok, thanks! I'll move the code to Java then 👍

robzienert commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProvider.java:
```diff
@@ -0,0 +1,70 @@
+    return cacheView.getAll(CLOUDFORMATION.getNs(), identifiers, RelationshipCacheFilter.none())
+      .stream()
+      .map(data -> {
+        log.debug(String.format("Cloud formation cached properties %s", data.getAttributes()));
```
You don't need to use `String.format` in log calls...

```java
log.debug("Cloud formation cached properties {}", data.getAttributes());
```

xavileon replied:
done in https://github.com/spinnaker/clouddriver/pull/3278/commits/85c6eadada2947091e4559d29078e29a08a40ba6

robzienert commented on clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonCloudFormationCachingAgentSpec.groovy:
```diff
@@ -0,0 +1,138 @@
+    1 * stackResults.stacks >> [ stack1, stack2 ]
+
+    results.find { it.id == Keys.getCloudFormationKey("stack1", "region", "account") }
+    results.find { it.id == Keys.getCloudFormationKey("stack2", "region", "account") }
```
Are you wanting to actually assert on the presence of these? I don't think these two lines don't do anything currently.

robzienert commented on clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonCloudFormationCachingAgentSpec.groovy:
```diff
@@ -0,0 +1,138 @@
+    def stackEvent = new StackEvent().withResourceStatus("CREATE_FAILED").withResourceStatusReason("who knows")
+    def stackEventResults = Mock(DescribeStackEventsResult)
+    def cache
+    def results
```
I'd just put the `def` of `cache` and `results` in the `when` stanza.

robzienert commented on clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProviderSpec.groovy:
```diff
@@ -0,0 +1,123 @@
+    "account1" || 2
+    "account2" || 1
+    "unknown"  || 0
+    null       || 0
```
Are you wanting to assert anything with `count`?

xavileon replied:
good question, it's a left over from a previous way of testing. I've changed the test a bit to make it more clear and explicit on the assertions.
done in https://github.com/spinnaker/clouddriver/pull/3278/commits/7455e0a53b9a12e6eed371e5cb7b9f4a895a7ec2

robzienert commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudFormation.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Map;
+
+/**
+ * A representation of a CloudDriver stack
```
`s/Driver/Formation/`

xavileon commented:
@robzienert thanks for the review 👍 addressed the comments about the logs in https://github.com/spinnaker/clouddriver/pull/3278/commits/85c6eadada2947091e4559d29078e29a08a40ba6 and the tests in https://github.com/spinnaker/clouddriver/pull/3278/commits/7455e0a53b9a12e6eed371e5cb7b9f4a895a7ec2

jervi commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonCloudFormationCachingAgent.java:
```diff
@@ -0,0 +1,118 @@
+import com.amazonaws.services.cloudformation.AmazonCloudFormation;
+import com.amazonaws.services.cloudformation.model.*;
+import com.amazonaws.services.cloudformation.model.Stack;
+import com.netflix.spinnaker.cats.agent.*;
```
Split into separate imports

jervi commented on clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonCloudFormationSpec.groovy:
```diff
@@ -0,0 +1,76 @@
+    cf.accountId == "accountId"
+    cf.stackStatus == "stackStatus"
+    cf.stackStatusReason == "stackStatusReason"
+  }
```
You can use the `with` method in spock:
```
    with(cf) {
      stackId == "stackId"
      tags == [tag1: "tag1", tag2: "tag2"]
      outputs == [out1: "out1", out2: "out2"]
      stackName == "stackName"
      region == "region"
      accountName == "accountName"
      accountId == "accountId"
      stackStatus == "stackStatus"
      stackStatusReason == "stackStatusReason"
    }
```

jervi commented on clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProviderSpec.groovy:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  @Unroll
+  void "list all cloud formations by account and region"() {
```
Tip: You can add variables to the name of unrolled tests:
`void "list all cloud formations by account '#account' and region '#region'"() {`
